### ISSUE TITLE
inconsistent autoscaling around axvlines

### ISSUE THREAD

**Issue by anntzer** [matplotlib/matplotlib#14651]
### Bug report

**Bug summary**

Autoscaling behavior around axvlines is inconsistent.

**Code for reproduction**

```python
from pylab import *

fig, axs = plt.subplots(2, 2)

axs[0, 0].axvline(0.9)

axs[0, 1].axvline(1.1)

axs[1, 0].plot([10, 11], [10, 11])
axs[1, 0].axvline(11.04)

axs[1, 1].plot([10, 11], [10, 11])
axs[1, 1].axvline(11.06)

plt.show()
```

**Actual outcome**

![test](https://user-images.githubusercontent.com/1322974/60385533-6ad4b500-9a8a-11e9-9c46-8ced09fd2735.png)

Top row: when the axvline is drawn in (0, 1) (the default), the margins are left as is; when the axvline is drawn outside of it, the margins move to surround it.

Bottom row: when the axvline is drawn within the margins imposed by the "other" line plot, the margins are left as is; when the axvline is drawn outside of it, yet another 5% margins are applied to it.

**Expected outcome**

Consistent behavior between left and right.

I think it's mostly just a matter of
```patch
--- i/lib/matplotlib/axes/_axes.py
+++ w/lib/matplotlib/axes/_axes.py
@@ -893,20 +886,13 @@ class Axes(_AxesBase):
 
         if "transform" in kwargs:
             raise ValueError(
-                "'transform' is not allowed as a kwarg;"
-                + "axvline generates its own transform.")
-        xmin, xmax = self.get_xbound()
-
-        # We need to strip away the units for comparison with
-        # non-unitized bounds
+                "'transform' is not allowed as a kwarg; "
+                "axvline generates its own transform")
         self._process_unit_info(xdata=x, kwargs=kwargs)
-        xx = self.convert_xunits(x)
-        scalex = (xx < xmin) or (xx > xmax)
-
         trans = self.get_xaxis_transform(which='grid')
         l = mlines.Line2D([x, x], [ymin, ymax], transform=trans, **kwargs)
         self.add_line(l)
-        self._request_autoscale_view(scalex=scalex, scaley=False)
+        self._request_autoscale_view(scalex=True, scaley=False)
         return l
 
     @docstring.dedent_interpd
```
(and likewise for axhline); OTOH I guess some may possibly actually be relying on non-expansion of margins when drawing axvlines just outside of the data limits but still within the margins?  dunno.

**Matplotlib version**

  * Operating system: linux
  * Matplotlib version: 3.1 
  * Matplotlib backend (`print(matplotlib.get_backend())`): qt5agg
  * Python version: 37
  * Jupyter version (if applicable):
  * Other libraries:

**Comment by timhoffm** [comment]
I guess autoscaling should always take avhlines into account.

**Comment by StevenLi-DS** [comment]
```python
fig, ax = plt.subplots()
ax.axvline(0.5, color='b', transform=ax.transAxes)
```
I got: `ValueError: 'transform' is not allowed as a kwarg;axvline generates its own transform.`

but the [doc](https://matplotlib.org/3.1.1/api/_as_gen/matplotlib.axes.Axes.axvline.html) suggested that `transform` is supported.

Am I missing something? If I wanna plot a line in the middle of a Axes, how should I do it for now?

**Comment by ImportanceOfBeingErnest** [comment]
@StevenLi-DS  `axvline` is a high-level convenience function for the special, but frequently used case where you want a vertical line spanning (part of the axes) positionned at a certain x coordinate in data units. The documentation **does not** suggest that `transform` is supported, in fact it clearly states

> Valid kwargs are Line2D properties, **with the exception of 'transform'**:

You can always achieve the same with a usual line and the respective transform, hence `ax.axvline(0.5)` is equivalent to `ax.plot([.5,.5], [0,1], transform=ax.get_xaxis_transform())`. 

If you want your line to be completely in axes coordinates, it would be `ax.plot([.5,.5], [0,1], transform=ax.transAxes)`. 

All of this is unrelated to the original issue here. It would help to use a new issue in such cases in order to not disturb the thread of arguments in existing ones.

**Comment by jklymak** [comment]
Back to the original thread.   I think a clone should not change the axes limits.  Ever.  It’s a way of annotating data, not data.  If someone wants to plot a vertical line then that’s what plot is for.

**Comment by ImportanceOfBeingErnest** [comment]
I think I would personally agree with "axvline should not change the axes limits". 

Related to this: https://github.com/matplotlib/matplotlib/issues/7742, which already did undergo two unsuccessful attempts to fix.

**Comment by anntzer** [comment]
So should it also not change limits even if plotting outside of the current limits?  (I'm fine with that solution too.)

**Comment by ImportanceOfBeingErnest** [comment]
That would be what *I* would expect. But maybe other users not? In particular, I think it would be a significant change to how it is handled now.

**Comment by jklymak** [comment]
In an ideal world it would not auto scale at all and explicitly say that in the docs

**Comment by anntzer** [comment]
It's probably not too hard to change that behavior with a deprecation warning (check whether the limits would change and if so warn), the only question is whether it's worth the trouble.

**Comment by timhoffm** [comment]
> axvline should not change the axes limits

not sure I agree with that. If the user plots this line, there's some intent behind it and it's most likely undesired if the line does not show up.

Example: Plot some relative values. Values larger than 1 represent data that got larger, values below 1 got smaller. It can be reasonable to `axhline(1)` to define the different semantics. It might be an essential information. If my data happen to be all smaller than 1, not taking into account the line would shift it off-screen, which would remove an important aspect of the graph.

**Comment by jklymak** [comment]
I'd argue in that case the user can specify those important limits manually.  My conceptualization of `axvline` is that its a special grid line.  

I appreciate other folks might conceptualize it differently, so we should decide if it autoscales (in x) or not, but not do something that is half way.

**Comment by dopplershift** [comment]
If I plot a vertical line at x=0.5, I'd expect 0.5 to be considered when automatically determining plot limits. I would not expect to need to manually tell matplotlib to include this.

**Comment by jklymak** [comment]
... and would you expect an auto scaling margin around the line?  If so, we should remove the funky logic and just autoscale.

**Comment by timhoffm** [comment]
I would expect a margin, for the same reason we have margins at all.

**Comment by anntzer** [comment]
@dopplershift If you plot a vertical line at x=0.99 while the current limits (computed on the rest of the data) is (0, 1), do you expect the 5% margin to go to 1.05(-epsilon)?
Just trying to clarify what everyone's opinion is...

**Comment by StevenLi-DS** [comment]
>The documentation does not suggest that transform is supported

The reason why I was confused is that `transform` is listed in Other Parameters:
![image](https://user-images.githubusercontent.com/33796896/63561938-bf942980-c529-11e9-8e85-d2382831992c.png)

https://matplotlib.org/3.1.1/api/_as_gen/matplotlib.axes.Axes.axvline.html****

**Comment by jklymak** [comment]
Yeah, thats an unfortunate artifact of just repeating the `Line2D` kwargs.  The caveat at the beginning tries to warn about that, but its kinda hard to catch if you are looking at a list of 30 properties.

**Comment by dopplershift** [comment]
@anntzer Personally, I expect the margin to increase as well. It may "just" be a notational line, but the whole point of margins is better aesthetics around things on the plot, keeping all items (dataspace wise) from being right on the edge.

**Comment by anntzer** [comment]
so in https://github.com/matplotlib/matplotlib/issues/14651#issue-462307767 you think we should increase the margins in the lower left plot?

**Comment by dopplershift** [comment]
Yes.

**Comment by jklymak** [comment]
Do we understand why the complicated logic was put in, and does removing it fail any tests?

**Comment by anntzer** [comment]
This came in in https://github.com/matplotlib/matplotlib/commit/d4965f (2008); all the commit message says is "fixed a problem with axh and vline scaling".  Previously it was never scaling, which is the behavior you (@jklymak) are advocating.

**Comment by jklymak** [comment]
I feel strongly that the behaviour should be simple and consistent.  I feel mildly that these should not participate in auto scaling, but don’t mind the opposite behaviour

**Comment by dopplershift** [comment]
I agree about simple and consistent. I can say that my reaction, if I call `axvline` or `axhline`, and the line does not appear in my plot due to unadjusted data bounds, will be WTF matplotlib?

**Comment by elmer-garduno** [comment]
pathces.Rectangle has the non-auto-scaling behavior, which is problematic if some things remain where they are and others are scaled. I agree the behavior should be consistent.

**Comment by timhoffm** [comment]
We should always autoscale because
- no autoscaling is bad as the artist may be off screen. - No Artists with explicitly given coordinates should have this behavior by default.
- partial autoscaling is weird and more complicated. Margins are there for a reason, namely to give relevant data a bit of room away from the Axes edge. That argument also holds for axvlines.

Note: The current behavior is even more awkward. The vline is always added to the view limits, only the autoscale update is conditionally suppressed, i.e. even if the line seems not to be considered initially, it will be picked up by a subsequent autoscale (which can happen explicitly through `autoscale_view()` but also as a side effect by adding more ar

```python
fig, (ax1, ax2, ax3) = plt.subplots(3)
ax1.plot([0, 1], [0, 1])
ax1.axvline(1.02)

ax2.plot([0, 1], [0, 1])
ax2.axvline(1.02)
ax2.autoscale_view()  # explicit autoscale view update

ax3.plot([0, 1], [0, 1])
ax3.axvline(1.02)
ax3.plot([0.2, 0.8], [0.5, 0.5])  # autoscale view update as a side effect
```

<img width="539" height="402" alt="Image" src="https://github.com/user-attachments/assets/dd185d19-0710-411b-9a1c-3cdf764925f9" />

The way forward is to always do the `_request_autoscale_view()` in https://github.com/matplotlib/matplotlib/blob/dd8d78b8dce60b6c8db86132892577a0b9dbe469/lib/matplotlib/axes/_axes.py#L895 and declare this as a bug fix.
